src/main/scala/stuff/products.scala

```diff
@@ -1,6 +1,90 @@
 package ohnosequences.stuff
 
 import functions._
+import products._
+
+abstract
+class Product {
+
+  type On <: Category
+  val  on  : Category.is[On]
+
+  @infix
+  type ×[X <: On#Objects, Y <: On#Objects] <: On#Objects
+
+  type ∗ <: On#Objects
+
+  def both[
+    X <: On#Objects,
+    A <: On#Objects,
+    B <: On#Objects,
+  ]
+  // TODO find a convenient aliasing convention for both ×'s
+  : ohnosequences.stuff.×[On#C[X, A], On#C[X, B]] -> On#C[X, A × B]
+
+  def erase[X <: On#Objects]: On#C[X, ∗]
+
+  def  left[A <: On#Objects, B <: On#Objects]: On#C[A × B, A]
+  def right[A <: On#Objects, B <: On#Objects]: On#C[A × B, B]
+}
+
+object Product {
+
+  type is[P <: Product] =
+    P {
+      type On = P#On
+      type ×[X <: P#On#Objects, Y <: P#On#Objects] = P# ×[X,Y]
+      type ∗ = P# ∗
+    }
+}
+
+// TODO make it final once we have assocs implemented
+abstract
+class CartesianMonoidalCategory[P <: Product](val product: Product.is[P]) {
+
+  type On = P#On
+  val on  = product.on
+
+  @infix
+  type ⊗[X <: On#Objects, Y <: On#Objects] = P # ×[X,Y]
+  type I = P # ∗
+
+  def ⊗[
+    A <: On#Objects,
+    B <: On#Objects,
+    C <: On#Objects,
+    D <: On#Objects
+  ]
+  : On#C[A,B] × On#C[C,D] -> On#C[A ⊗ C, B ⊗ D] =
+    λ { fg =>
+      product.both(
+        on.composition( product.left[A,C] and fg.left ) and
+        on.composition(product.right[A,C] and fg.right)
+      )
+    }
+
+  // TODO these two are nice exercises
```

Comment:
What are exactly the exercises? I am not sure I understand how those two methods should behave.

Reply:
Implementing these two associators in terms of the other methods.

Reply:
Yes, but what should they do? It seems like it's something trivial, but I don't see it 😞

Reply:
`(A × B) × C` and `A × (B × C)` are *different*, though isomorphic, types. These two associators (should) represent this isomorphism. You can implement them through repeated use of `left`, `right`, and `both`. Draw it first!